src/Open3D/Visualization/Rendering/RendererHandle.h

```diff
@@ -0,0 +1,194 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include <array>
+#include <cstdint>
+#include <functional>
+
+#define FMT_HEADER_ONLY 1
+#define FMT_STRING_ALIAS 1
+// Including windows.h causes all kinds of #defines like "OPAQUE", "near",
+// "far", causes compile errors with Filament includes, and generally wrecks
+// havoc.
+#ifndef FMT_USE_WINDOWS_H
+#define FMT_USE_WINDOWS_H 0
+#endif
+#include <fmt/format.h>
+
+namespace open3d {
+
+namespace visualization {
+
+// If you add entry here, don't forget to update TypeToString!
+enum class EntityType : std::uint16_t {
+    None = 0,
+
+    View,
+    Scene,
+
+    Geometry,
+    Light,
+    IndirectLight,
+    Skybox,
+    Camera,
+    Material,
+    MaterialInstance,
+    Texture,
+
+    VertexBuffer,
+    IndexBuffer,
+
+    Count
+};
+
+// RenderEntityHandle - handle type for entities inside Renderer
+// Can be used in STL containers as key
+struct REHandle_abstract {
+    static const char* TypeToString(EntityType type);
+
+    static const std::uint16_t kBadId = 0;
+    const EntityType type = EntityType::None;
+
+    inline size_t Hash() const {
+        return static_cast<std::uint16_t>(type) << 16 | id;
+    }
+
+    bool operator==(const REHandle_abstract& other) const {
+        return id == other.id && type == other.type;
+    }
+
+    bool operator<(const REHandle_abstract& other) const {
+        return Hash() < other.Hash();
+    }
+
+    explicit operator bool() const { return id != kBadId; }
+
+    REHandle_abstract() : type(EntityType::None), id(kBadId) {}
+
+    std::uint16_t GetId() const { return id; }
+
+protected:
+    REHandle_abstract(const EntityType aType, const std::uint16_t aId)
+        : type(aType), id(aId) {}
+
+    static std::array<std::uint16_t, static_cast<size_t>(EntityType::Count)>
+            uid_table;
+
+    std::uint16_t id = kBadId;
+};
+
+std::ostream& operator<<(std::ostream& os, const REHandle_abstract& uid);
+
+// REHandle is used for specification of handle types to prevent
+// errors with passing, assigning or comparison of different kinds of handles
+template <EntityType entityType>
+struct REHandle : public REHandle_abstract {
+    static const REHandle kBad;
+
+    static REHandle Next() {
+        const auto index = static_cast<std::uint16_t>(entityType);
+        auto id = ++uid_table[index];
+        if (id == REHandle_abstract::kBadId) {
+            uid_table[index] = REHandle_abstract::kBadId + 1;
+            id = REHandle_abstract::kBadId + 1;
+        }
+
+        return std::move(REHandle(id));
+    }
+
+    static REHandle Concretize(const REHandle_abstract& abstract) {
+        if (abstract.type != entityType) {
+            // assert("Incompatible render uid types!\n");
```

Comment:
Probably? You know as much as I do about this code. I'm not a fan of aborting with LogError, especially since a bad handle is just going to result in an incorrect image, not a corrupted file (like if this were are word processor) or something.